pkg/monitortests/apps/pdbunhealthypodevictionpolicy/monitortest.go

```diff
@@ -0,0 +1,123 @@
+package pdbunhealthypodevictionpolicy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	"github.com/openshift/origin/pkg/monitortestframework"
+	"github.com/openshift/origin/pkg/test/ginkgo/junitapi"
+	policyv1 "k8s.io/api/policy/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var policyExceptions = map[string]sets.Set[string]{
+	"openshift-etcd": sets.New("etcd-guard-pdb"),
+}
+
+type pdbUnhealthyPodEvictionPolicyChecker struct {
+	kubeClient kubernetes.Interface
+}
+
+func NewAnalyzer() monitortestframework.MonitorTest {
+	return &pdbUnhealthyPodEvictionPolicyChecker{}
+}
+
+func (w *pdbUnhealthyPodEvictionPolicyChecker) StartCollection(ctx context.Context, adminRESTConfig *rest.Config, recorder monitorapi.RecorderWriter) error {
+	var err error
+	w.kubeClient, err = kubernetes.NewForConfig(adminRESTConfig)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *pdbUnhealthyPodEvictionPolicyChecker) CollectData(ctx context.Context, storageDir string, beginning, end time.Time) (monitorapi.Intervals, []*junitapi.JUnitTestCase, error) {
+	if w.kubeClient == nil {
+		return nil, nil, nil
+	}
+
+	namespaces, err := w.kubeClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	junits := []*junitapi.JUnitTestCase{}
+	for _, ns := range namespaces.Items {
+		// ignore openshift-must-gather-* namespaces which are generated dynamically
+		isPermanentOpenShiftNamespace := (ns.Name == "openshift" || strings.HasPrefix(ns.Name, "openshift-")) && !strings.HasPrefix(ns.Name, "openshift-must-gather-")
+		if !isPermanentOpenShiftNamespace {
+			continue
+		}
+
+		pdbs, err := w.kubeClient.PolicyV1().PodDisruptionBudgets(ns.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		failures := make([]string, 0)
+		for _, pdb := range pdbs.Items {
+			if pdb.Spec.UnhealthyPodEvictionPolicy != nil && *pdb.Spec.UnhealthyPodEvictionPolicy == policyv1.AlwaysAllow {
+				continue
+			}
+			if policyException, ok := policyExceptions[pdb.Namespace]; ok && policyException.Has(pdb.Name) {
+				continue
+			}
+
+			failures = append(failures, fmt.Sprintf("AlwaysAllow .spec.unhealthyPodEvictionPolicy should be set on a %q PDB, found %q policy instead.", pdb.Name, ptr.Deref(pdb.Spec.UnhealthyPodEvictionPolicy, "<nil>")))
+		}
+
+		testName := fmt.Sprintf("[sig-apps] all PodDisruptionBudgets in ns/%s should have a %q .spec.unhealthyPodEvictionPolicy", ns.Name, policyv1.AlwaysAllow)
+		if len(failures) == 0 {
+			junits = append(junits, &junitapi.JUnitTestCase{Name: testName})
+			continue
+		}
+
+		failureMsg := fmt.Sprintf("%v\n\n%v", strings.Join(failures, "\n"), PDBAlwaysAllowPolicyExplanation())
```

Comment:
I think you want this in `EvaluateTestsFromConstructedIntervals`.

That tends to gather all the intervals and then fail the test at the end if you detect a bad interval.